darshan-util/pydarshan/darshan/experimental/plots/plot_dxt_heatmap.py

```diff
@@ -0,0 +1,124 @@
+# -*- coding: utf-8 -*-
+
+from darshan.report import *
+
+import matplotlib
+import matplotlib.pyplot as plt
+from mpl_toolkits.axes_grid1 import make_axes_locatable 
+import numpy as np
+
+def plot_dxt_heatmap(report, 
+        xbins=10, ybins=None, 
+        group_by="rank", 
+        mods=None, ops=None, 
+        display_values=False, cmap=None, figsize=None, 
+        amplify=False):
+    """
+    Generates a heatmap plot from a report with DXT traces.
+
+    Args:
+        report (darshan.DarshanReport): report to generate plot from
+        xbins (int): number of bins on the x axis
+        ybins (int): number of bins on the y axis
+        group_by (str): attribute to group by (e.g., rank, hostname)
+        mods (list): modules to include in heatmap (e.g., ['DXT_POSIX', 'DXT_MPIIO'])
+        ops (list): operations to consider (e.g., ['read', 'write']
+        display_values (bool): show values per heatmap field
+        cmap: overwrite colormap (see matplotlib colormaps)
+        figsize: change figure size (see matplotlib figsize)
+
+        amplify (int): paint neighbouring cells e.g., when working with many ranks
+    """
+
+    runtime = report.end_time - report.start_time       # timedelta
+    runtime = runtime.seconds + runtime.microseconds    # to float/int
+    if runtime == 0:
+        runtime = 1
+
+        
+    if mods is None:
+        mods = ['DXT_POSIX', 'DXT_MPIIO']
+    
+    if ops is None:
+        ops = ['read', 'write']
+    
+    
+    def get_hostname_mapper(report):
+        """ Determine which ranks map to which hostname. """
+        hostnames = {}
+        i = 0
+        for mod in mods:
+            for rec in report.records[mod]:
+                hostname = rec['hostname']
+                if hostname not in hostnames:
+                    hostnames[hostname] = i
+                    i += 1
+        return hostnames
+   
+
+    hostnames = get_hostname_mapper(report)
+    ranks = report.metadata['job']['nprocs']
+    
+    # set ybins, yticks and ylabel to match choice of group_by 
+    if group_by in ['hostname', 'node']:
+        ybins = len(hostnames)
+        yticks = list(hostnames.keys())
+        ylabel = "Hostname"
+    else:
+        if ybins is None:
+            ybins = ranks 
+            ylabel = "Rank"
+            #yticks = [str(x) for x in range(ybins)] 
+        else:
+            ylabel = f"Binned Ranks (binsize={int(ranks/ybins)} ranks)"
+
+    
+    # heatmap to be populated with event counts
+    events = np.zeros((ybins, xbins))    
+    
+    for mod in mods:
+        for op in ops:
+            for rec in report.records[mod]:
+                
+                if group_by in ['hostname', 'node']:
+                    ybin = hostnames[rec['hostname']]
+                else:
+                    ybin = int((rec['rank'] /ranks) * ybins)
+                    
+                    
+                for event in rec[f'{op}_segments']:
```

Comment:
Here we are already nested 4 levels deep in a `for` loop, iterating over a `list` of dictionary objects representing individual read or write events. We're perhaps somewhat hamstrung by the generality of `DarshanRecordCollection`, which as its comments suggest has to handle a diverse set of nested data structures for all the instrumentation modules.

Even with more than 2k ranks performance doesn't really seem to be an issue here in my hands though. So, I guess I'm just saying that it is perhaps a bit unfortunate that we can't leverage i.e., a specialized `pandas` dataframe object for each type of darshan instrumentation module, but I'm also not sure it matters too much for the present task.

It is true that we could use `to_df()`, but we're ultimately still dealing with a collection of `DarshanRecordCollection` values which happen to have some internal dataframes in that case, which still leaves us with awkward pure-Python looping instead of vectorized ops in `pandas` or a similar framework.

Reply:
Longer-term, I'm not sure if we need to think about specializing/structuring the exposed DXT data some other way. One challenge might be that the data is (potetenially) "ragged" to some extent as I understand it. So you might have a dataframe a bit like this if you went that route:

```
rank, write_count, write_segment_1_start_time, write_segment_2_start_time, ...
0           1                  0.1                             NaN
1           2                  0.2                             0.4
```

Where you might end up with many rows/ranks filled with `NaN`s, or decide to store a collection in the dataframe, which may bring you back to the same place of nested data structures again.

Reply:
I as well would update the DXT representation because currently it very closely mirrors the raw representation in the log. As there is not really a lot of experience on a "most common mode of consumption" that seemed to be sensible for a start, but I agree about it being awkward to access and handle as is.
I find myself often needing read and write activity combined in a sorted list by start_time, for example. 

So actually I am also very much in favor of having all tabular style record types using pandas data frames or xarray internally and rely on all the community optimizations and user familiarity of their APIs in all sorts of analysis use cases.
I did some performance testing with converting everything to pandas on import in the CFFI layer: Currently this does not perform unfortunately, but that can be solved by adding a log_get_n_records on the C side ;)   Another related thing we wanted to address in combination is looking at ways to allow seeking across modules/(maybe records), which due to the sequence of records being compressed requires a little oversight to not break the existing utils.

I think introducing specialized record collections is a good idea, and I do think the current RecordCollection is exactly in that spirit, just that it tries to find some common functionality that helps with the export to_json/to_dict/do_df or merging/filtering when dealing with multiple reports. In principle I would image to_df in many cases to become a noop that just returns the internal pandas data frame. 

From my perspective this is a trade-off in any case that requires to decide how much one would like to preserve Darshans conceptional and raw data model of breaking multi-level instrumentation into <darshan-log>/<module/layer>/records[].